apps/sms/js/startup.js

```diff
@@ -112,17 +112,19 @@ var Startup = {
       MessageManager.init();
       Navigation.init();
 
-      // If target panel is different from the default one, let's load remaining
-      // scripts as soon as possible, otherwise we can wait until first page of
-      // threads is loaded and rendered.
-      var panelName = Navigation.getPanelName();
-      if (panelName && panelName !== 'thread-list') {
-        // Initialize default panel only after target panel is ready.
+      var panelName = Navigation.getPanelName() || Navigation.defaultPanel;
+
+      // If initial panel is default one and app isn't run from notification,
+      // then just navigate to it, otherwise we can delay default panel
+      // initialization until we navigate to requested non-default panel.
+      if (panelName === Navigation.defaultPanel &&
+        !navigator.mozHasPendingMessage('notification')) {
+        Navigation.toPanel(panelName);
```

Comment:
grrr I'd like to use mozHasPendingMessage without an argument but it's not possible :( For example we probably want to do the same with sms-received so that we get the notification earlier... Not as important though but this is not a really good code :/

So let me summarize.

We want to move to the defaultPanel only when:
* there is no message (means no hash (getPanelName() returns the empty string) and no pending message for 'notification')
* no other panel has been requested

Here in this code I don't like that we use a lot of functions from Navigation, I feel like some or most of this code should be in Navigation.

So I'd keep the existing toPanelFromHash (but I'd still remove the call from Navigation.init) and do something like:
```javascript
if (Navigation.isDefaultPanel() && !navigator.mozHasPendingMessage('notification')) {
  Navigation.toDefaultPanel(); // here this could return a promise, and we could put the setReady/fullyLoaded mark as handler as this promise (same for toPanelFromHash below)
  initializeDefaultPanel(...); // maybe the content of this could be moved to `InboxView.beforeEnter()` (without returning the promise) ?
} else {
  Navigation.toPanelFromHash(); // modified to not move to default panel if there is no hash
  Navigation.on('navigated', ...); // here the handler could be simply InboxView.beforeEnter()
}
```

If I change according to my comments, here is how this could be:

```javascript
  // assuming toPanelFromHash would reject the promise if we can't handle the hash
var  panelPromise = Navigation.toPanelFromHash().then(
  function onLoaded() {
    Navigation.on('navigated', InboxView.beforeEnter.bind(InboxView));
    finishLazyload();
  },
  function onNotLoaded() {
    if (!navigator.mozHasPendingMessage('notification') {
      return Navigation.toDefaultPanel();
    }
    // else do nothing, notifications will be handled later in activity_handler
  }
).then(() => {
  setReady();
  markFullyLoaded();
});
```

Not sure this is better, maybe it would be more readable by exporting to external functions. But tell me what you think !